File: service/openark/api/v1/skills.py
from fastapi import APIRouter, Depends, Request

from ...core.models import (
    DistillRequest,
    DistillResponse,
    SkillsResponse,
    SkillSummary,
    SkillVerifyResponse,
)
from ...core.registry import AgentRegistry
from ...modules.skills import SkillsModule
from .agents import get_registry, require_agent

router = APIRouter(tags=["skills"])


def get_skills_module(request: Request) -> SkillsModule:
    return request.app.state.modules.skills


@router.get("/agents/{name}/skills", response_model=SkillsResponse)
def list_skills(
    name: str,
    drafts: bool = False,
    registry: AgentRegistry = Depends(get_registry),
    module=Depends(get_skills_module),
):
    agent_home = require_agent(registry, name)
    skills = module.list(agent_home, include_drafts=drafts)
    return SkillsResponse(
        skills=[
            SkillSummary(name=s.name, description=s.description, status=s.status) for s in skills
        ]
    )


@router.post("/agents/{name}/skills/distill", response_model=DistillResponse)
def distill_skill(
    name: str,
    request: DistillRequest,
    registry: AgentRegistry = Depends(get_registry),
    module=Depends(get_skills_module),
):
    require_agent(registry, name)
    return module.distill(registry, name, request.trace)


@router.post("/agents/{name}/skills/{slug}/verify", response_model=SkillVerifyResponse)
def verify_skill(
    name: str,
    slug: str,
    registry: AgentRegistry = Depends(get_registry),
    module=Depends(get_skills_module),
):
    require_agent(registry, name)
    return module.verify(registry, name, slug)

File: plugin/src/modules/skills.ts
import { z } from "zod";
import { requiredString } from "../core/args.js";
import { openarkHome } from "../core/config.js";
import { linkSkills, opencodeConfigDir } from "../core/installer.js";
import type { InjectionBlock, ModuleContext, ModuleTool, OpenArkModule } from "../core/types.js";
import type { components } from "../generated/api-types.js";

type SkillsResponse = components["schemas"]["SkillsResponse"];
type DistillResponse = components["schemas"]["DistillResponse"];
type SkillVerifyResponse = components["schemas"]["SkillVerifyResponse"];
type Skill = components["schemas"]["SkillSummary"];

const skillDistillShape = { trace: requiredString("trace is required") };
const skillDistillArgs = z.object(skillDistillShape);
const skillVerifyShape = { name: requiredString("name is required") };
const skillVerifyArgs = z.object(skillVerifyShape);
const skillsListShape = { all: z.boolean().optional() };

export const skillsModule: OpenArkModule = {
  name: "skills",
  description: "Distilled, reusable procedures materialized as opencode skills",

  async init(ctx) {
    if (!(await ctx.service.health())) {
      throw new Error("openark service unreachable");
    }
  },

  async injections(ctx: ModuleContext): Promise<InjectionBlock[]> {
    try {
      const res = await ctx.service.getJSON<SkillsResponse>(`/v1/agents/${ctx.agent}/skills`);
      const skills = (res.skills ?? []).filter((s: Skill) => s.name?.trim());
      if (!skills.length) return [];
      const lines = skills.map((s: Skill) => `- ${s.name}: ${s.description}`).join("\n");
      return [
        {
          title: "Learned skills available",
          body: `These skills exist as opencode skills; prefer them over reinventing:\n${lines}`,
          priority: 50,
        },
      ];
    } catch {
      return [];
    }
  },

  tools(ctx: ModuleContext): ModuleTool[] {
    return [
      {
        name: "skill_distill",
        description:
          "Distill a successfully completed multi-step workflow into a reusable draft skill. " +
          "Pass the workflow trace (what you did, in order) as the trace argument.",
        argsSchema: skillDistillShape,
        execute: async (args) => {
          const { trace } = skillDistillArgs.parse(args);
          return ctx.service.postJSON<DistillResponse>(`/v1/agents/${ctx.agent}/skills/distill`, {
            trace,
          });
        },
      },
      {
        name: "skills_list",
        description: "List learned skills (verified by default; pass all: true to include drafts)",
        argsSchema: skillsListShape,
        execute: async (args) => {
          const includeDrafts = args.all === true;
          return ctx.service.getJSON<SkillsResponse>(
            `/v1/agents/${ctx.agent}/skills?drafts=${includeDrafts ? "true" : "false"}`,
          );
        },
      },
      {
        name: "skill_verify",
        description:
          "Verify a draft skill by name, promoting it so it is advertised and materialized " +
          "as an opencode skill. Use after one successful reuse or explicit user approval.",
        argsSchema: skillVerifyShape,
        execute: async (args) => {
          const { name: slug } = skillVerifyArgs.parse(args);
          const result = await ctx.service.postJSON<SkillVerifyResponse>(
            `/v1/agents/${ctx.agent}/skills/${encodeURIComponent(slug)}/verify`,
            {},
          );
          if (result.verified) {
            // Materialize immediately so the next injection's promise ("this
            // exists as an opencode skill") is true without an `openark
            // install` round-trip.
            try {
              linkSkills(opencodeConfigDir(), openarkHome());
            } catch (err) {
              ctx.log("warn", `skill link after verify failed: ${String(err)}`);
            }
          }
          return result;
        },
      },
    ];
  },
};
